src/celeritas/global/ActionLauncher.device.hh

```diff
@@ -67,25 +84,132 @@ class ActionLauncher
                   "Launched action must be a trivially copyable function "
                   "object");
 
+    // Function pointer types to specify the correct overloaded kernel
+    using KernelNumThreads = void (*)(size_type const, F);
+
+    using KernelThreadRange = void (*)(Range<ThreadId> const, F);
+
+    using KernelFunction
+        = std::function<void(CoreState<MemSpace::device> const&, F const&)>;
+
   public:
     //! Create a launcher from an action
     explicit ActionLauncher(ExplicitActionInterface const& action)
-        : calc_launch_params_{action.label(), &launch_action_impl<F>}
+        : calc_launch_params_{init_kernel_param_calculator(action.label(),
+                                                           false)}
+        , action_id_{action.action_id()}
+        , invoke_kernel_{bind_invoke_kernel(false)}
     {
     }
 
     //! Create a launcher with a string extension
     ActionLauncher(ExplicitActionInterface const& action, std::string_view ext)
-        : calc_launch_params_{action.label() + "-" + std::string(ext),
-                              &launch_action_impl<F>}
+        : calc_launch_params_{init_kernel_param_calculator(
+            action.label() + "-" + std::string(ext), false)}
+        , action_id_{action.action_id()}
+        , invoke_kernel_{bind_invoke_kernel(false)}
+    {
+    }
+
+    //! Create a launcher from an action, minimizing kernel grid size if
+    //! sorting is done for this action
+    explicit ActionLauncher(CoreParams const& params,
+                            ExplicitActionInterface const& action)
+        : ActionLauncher(
+            params,
+            action,
+            action.label(),
+            action.order() == ActionOrder::post
+                    && params.init()->host_ref().track_order
+                           == TrackOrder::sort_step_limit_action
+                || action.order() == ActionOrder::along
+                       && params.init()->host_ref().track_order
+                              == TrackOrder::sort_along_step_action)
+    {
+    }
+
+    //! Create a launcher with a string extension, minimizing kernel grid size
+    //! if sorting is done for this action
+    ActionLauncher(CoreParams const& params,
```

Comment:
Params are effectively immutable, but unlike the `static` launch helper, they're not global: we can (and in unit tests *do*) use different params instances in the same execution.

But ugh that makes me realize there's another bug lurking: we store the action ID during the static constructor, but the ID could change for different instances of Params. I guess it would be better to move away entirely from static variables...

Regarding the kernel doubling: I'm guessing the compiler spends a fair bit of time optimizing each kernel independently, so it seems likely the time spent will increase greatly.

Reply:
Mmh, I'd expect the launcher to be bound to the action it's given in the constructor since it's instantiated in the said action. So you mean that the action id can be changed after an action is created?

Reply:
I'm passing the action again in `operator()` to make sure we use the correct action id, however we still use it in the constructor to register the action to the kernel registry.